Occasional crash related to OID handling

Unit tests fail occasionally in CI, reproducable with a large `-count` value either with a crash:
```
--- FAIL: TestGssOidSetFromOids (0.00s)
panic: runtime error: cgo argument has Go pointer to unpinned Go pointer [recovered, repanicked]
```

 or a corrupt OID:
```
-- FAIL: TestGssOidSetFromOids (0.00s)
    common_test.go:46:
                Error Trace:    /Users/jacob/src/github.com/golang-auth/go-gssapi-c/common_test.go:46
                                                        /Users/jacob/src/github.com/golang-auth/go-gssapi-c/helpers_test.go:39
                Error:          Received unexpected error:
                                asn1: syntax error: zero length OBJECT IDENTIFIER
                Test:           TestGssOidSetFromOids
    common_test.go:48: Stopping test TestGssOidSetFromOids due to fatal error

```

--- a/.github/workflows/full-tests.yml
+++ b/.github/workflows/full-tests.yml
@@ -106,6 +106,7 @@ jobs:
             PKGS="${PKGS} heimdal"
           fi
 
+          echo Installing packages: ${PKGS}
           pkg install -y ${PKGS}
 
         run: |

--- a/.gitignore
+++ b/.gitignore
@@ -24,3 +24,4 @@ toolbin/
 
 coverage.html
 go-gssapi-c.test
+*.code-workspace

--- a/Makefile
+++ b/Makefile
@@ -1,9 +1,8 @@
 current_dir := $(dir $(abspath $(lastword $(MAKEFILE_LIST))))
 GO          ?= go
-TOOLBIN := $(current_dir)/toolbin
-
-src_dir = $(current_dir)
-
+GOOS 		?= $(shell $(GO) env GOOS)
+GOARCH 		?= $(shell $(GO) env GOARCH)
+TOOLBIN 	:= $(current_dir)/toolbin/$(GOOS)_$(GOARCH)
 
 .DEFAULT: build
 
@@ -12,38 +11,40 @@ build: generate
 	./scripts/gofmt
 
 .PHONY: generate
-generate: $(src_dir)/testvecs_test.go
+generate: testvecs_test.go
 
-$(src_dir)/testvecs_test.go: build-tools/mk-test-vectors
-	$(GO) generate $(src_dir)/common_test.go
+testvecs_test.go: build-tools/mk-test-vectors
+	$(GO) generate common_test.go
 
+PKGS = $(shell $(GO) list ./... | egrep -v '/examples/|/build-tools/')
 
 .PHONY: test
-test:
-	./scripts/gofmt
-	${GO} test -asan ./... -coverprofile=cover.out -covermode=atomic
-	${GO} tool cover -html=cover.out -o coverage.html
+test: $(TOOLBIN)/gocovmerge $(TOOLBIN)/go-test-coverage
+	@echo "==> check code formatting"
+	@./scripts/gofmt
+	@echo "==> run tests for " $(PKGS)
+	@${GO} test $(PKGS) -coverprofile=cover.out -covermode=atomic
+	@go tool cover -html=cover.out -o coverage.html
+	@$(TOOLBIN)/go-test-coverage --config .testcoverage.yml
+
+.PHONY: lint
+lint: | $(TOOLBIN)/golangci-lint
+	$(TOOLBIN)/golangci-lint run 
+
+.PHONY: tools
+tools: $(TOOLBIN)/golangci-lint $(TOOLBIN)/gocovmerge $(TOOLBIN)/go-test-coverage
+	@echo "==> installing required tooling..."
+
+$(TOOLBIN)/golangci-lint:
+	GOBIN=$(TOOLBIN) GO111MODULE=on $(GO) install github.com/golangci/golangci-lint/cmd/golangci-lint@v1
+
+$(TOOLBIN)/gocovmerge:
+	GOBIN=$(TOOLBIN) GO111MODULE=on $(GO) install github.com/wadey/gocovmerge@latest
+
+$(TOOLBIN)/go-test-coverage:
+	GOBIN=$(TOOLBIN) GO111MODULE=on $(GO) install github.com/vladopajic/go-test-coverage/v2@latest
 
 .PHONY: gdb
 gdb:
 	GO_CFLAGS="-g" $(GO) test -c  -timeout 30s
 	gdb ./go-gssapi-c.test
-
-.PHONY: lint
-lint: | $(TOOLBIN)/golangci-lint $(TOOLBIN)/staticcheck
-	@echo ------ golangci-lint
-	@echo ------
-	$(TOOLBIN)/golangci-lint run
-	@echo -e "\n------ staticcheck"
-	@echo ------
-	$(TOOLBIN)/staticcheck 
-
-.PHONY: tools
-tools: $(TOOLBIN)/golangci-lint $(TOOLBIN)/staticcheck
-	@echo "==> installing required tooling..."
-
-$(TOOLBIN)/golangci-lint: | $(GOENV)
-	GOBIN=$(TOOLBIN) GO111MODULE=on $(GO) install github.com/golangci/golangci-lint/v2/cmd/golangci-lint@v2
-
-$(TOOLBIN)/staticcheck: ~$(GOENV)
-	GOBIN=$(TOOLBIN) GO111MODULE=on $(GO) install honnef.co/go/tools/cmd/staticcheck@2025.1.1

--- a/contrib/pinning/go.mod
+++ b/contrib/pinning/go.mod
@@ -1,6 +1,6 @@
 module test
 
-go 1.25.1
+go 1.24
 
 require github.com/stretchr/testify v1.11.1
 

--- a/cred.go
+++ b/cred.go
@@ -31,8 +31,11 @@ func hasDuplicateCred() bool {
 
 func (provider) AcquireCredential(name g.GssName, mechs []g.GssMech, usage g.CredUsage, lifetime *g.GssLifetime) (g.Credential, error) {
 	// turn the mechs into an array of OIDs
-	cOidSet, pinner := gssOidSetFromOids(mechsToOids(mechs), nil)
-	defer pinner.Unpin()
+	cOidSet, err := newOidSet(mechsToOids(mechs))
+	if err != nil {
+		return nil, err
+	}
+	defer cOidSet.Release() //nolint:errcheck
 
 	var cGssName C.gss_name_t = C.GSS_C_NO_NAME
 	if name != nil {
@@ -46,7 +49,7 @@ func (provider) AcquireCredential(name g.GssName, mechs []g.GssMech, usage g.Cre
 
 	var minor C.OM_uint32
 	var cCredID C.gss_cred_id_t = C.GSS_C_NO_CREDENTIAL
-	major := C.gss_acquire_cred(&minor, cGssName, gssLifetimeToSeconds(lifetime), cOidSet, C.int(usage), &cCredID, nil, nil)
+	major := C.gss_acquire_cred(&minor, cGssName, gssLifetimeToSeconds(lifetime), cOidSet.oidSet, C.int(usage), &cCredID, nil, nil)
 
 	if major != 0 {
 		return nil, makeStatus(major, minor)

--- a/helpers.go
+++ b/helpers.go
@@ -52,38 +52,6 @@ func oidsFromGssOidSet(oidSet C.gss_OID_set) []g.Oid {
 	return ret
 }
 
-// Convert from a slice of OID object to a set of C OIDS (see below)
-func gssOidSetFromOids(oids []g.Oid, pinner *runtime.Pinner) (C.gss_OID_set, *runtime.Pinner) {
-	if pinner == nil {
-		pinner = &runtime.Pinner{}
-	}
-
-	if len(oids) == 0 {
-		return nil, pinner
-	}
-
-	// Create a Go slice containing the C gss_OID elements, which is gauranteed to be
-	// contiguous, and suitable for use as a C array
-	oidPtrs := make([]C.gss_OID, len(oids))
-
-	// All the OID pointers will be pinned after this loop
-	for i, oid := range oids {
-		// oid2Coid will pin memory pointed to by gss_OID.elements
-		oidPtrs[i], pinner = oid2Coid(oid, pinner)
-
-		// but we also need to pin the gss_OID struct itself because its a member
-		// of gss_OID_set
-		pinner.Pin(oidPtrs[i])
-	}
-
-	cOidSet := C.gss_OID_set_desc{
-		count:    C.size_t(len(oids)),
-		elements: oidPtrs[0],
-	}
-
-	return &cOidSet, pinner
-}
-
 // Convert Go GSSAPI interface mech names to a set of mech OIDs
 func mechsToOids(mechs []g.GssMech) []g.Oid {
 	ret := make([]g.Oid, len(mechs))

--- a/helpers_mac.go
+++ b/helpers_mac.go
@@ -45,6 +45,8 @@ func oid2Coid(oid g.Oid, pinner *runtime.Pinner) (C.gss_OID, *runtime.Pinner) {
 	if len(oid) > 0 {
 		pinner.Pin(&oid[0])
 		var cOid C.gss_OID_desc
+
+		// Go doesn't know about the elements field of the gss_OID_desc struct due to alignment issues
 		C._set_oid_fields(&cOid, C.OM_uint32(len(oid)), unsafe.Pointer(&oid[0]))
 		return &cOid, pinner
 	} else {

--- a/helpers_test.go
+++ b/helpers_test.go
@@ -2,48 +2,9 @@ package gssapi
 
 import (
 	"testing"
-	"unsafe"
 
 	g "github.com/golang-auth/go-gssapi/v3"
 )
-
-func TestGssOidSetFromOids(t *testing.T) {
-	assert := NewAssert(t)
-
-	oids := []g.Oid{}
-
-	// test with no OIDs
-	oidSet, pinner := gssOidSetFromOids(oids, nil)
-	assert.NotNil(pinner)
-	if pinner != nil {
-		pinner.Unpin()
-	}
-	assert.Nil(oidSet)
-
-	mechs := []g.GssMech{g.GSS_MECH_KRB5, g.GSS_MECH_IAKERB}
-
-	// Test with some OIDs
-	oids = append(oids, mechs[0].Oid(), mechs[1].Oid())
-	oidSet, pinner = gssOidSetFromOids(oids, nil)
-	assert.NotNil(oidSet)
-	assert.NotNil(pinner)
-
-	if oidSet != nil {
-		assert.Equal(2, int(oidSet.count))
-
-		elms := oidSet.elements
-		s := unsafe.Slice(elms, oidSet.count)
-		for i, cOid := range s {
-			oid := oidFromGssOid(&cOid)
-			oidString, err := oid2String(oid)
-			assert.NoErrorFatal(err)
-			assert.Equal(mechs[i].OidString(), oidString)
-		}
-	}
-	if pinner != nil {
-		pinner.Unpin()
-	}
-}
 
 func TestOid2String(t *testing.T) {
 	assert := NewAssert(t)

--- a/helpers_unix.go
+++ b/helpers_unix.go
@@ -17,6 +17,10 @@ import (
 import "C"
 
 func oidFromGssOid(cOid C.gss_OID) g.Oid {
+	if cOid == C.GSS_C_NO_OID {
+		return nil
+	}
+
 	return C.GoBytes(cOid.elements, C.int(cOid.length))
 }
 
@@ -30,10 +34,12 @@ func oid2Coid(oid g.Oid, pinner *runtime.Pinner) (C.gss_OID, *runtime.Pinner) {
 	}
 
 	if len(oid) > 0 {
+		p := unsafe.Pointer(&oid[0])
 		pinner.Pin(&oid[0])
+
 		return &C.gss_OID_desc{
 			length:   C.OM_uint32(len(oid)),
-			elements: unsafe.Pointer(&oid[0]),
+			elements: p,
 		}, pinner
 	} else {
 		return C.GSS_C_NO_OID, pinner

--- a/oidset.go
+++ b/oidset.go
@@ -1,0 +1,58 @@
+package gssapi
+
+import (
+	"runtime"
+
+	g "github.com/golang-auth/go-gssapi/v3"
+)
+
+/*
+#include "gss.h"
+*/
+import "C"
+
+type oidSet struct {
+	pinner *runtime.Pinner
+	oidSet C.gss_OID_set
+}
+
+func newOidSet(oids []g.Oid) (*oidSet, error) {
+	ret := &oidSet{
+		pinner: &runtime.Pinner{},
+		oidSet: nil,
+	}
+
+	if len(oids) == 0 {
+		return ret, nil
+	}
+
+	var cMinor C.OM_uint32
+	cMajor := C.gss_create_empty_oid_set(&cMinor, &ret.oidSet)
+	if cMajor != C.GSS_S_COMPLETE {
+		return nil, makeStatus(cMajor, cMinor)
+	}
+
+	for _, oid := range oids {
+		cOid, _ := oid2Coid(oid, ret.pinner)
+		cMajor := C.gss_add_oid_set_member(&cMinor, cOid, &ret.oidSet)
+		if cMajor != C.GSS_S_COMPLETE {
+			return nil, makeStatus(cMajor, cMinor)
+		}
+	}
+
+	return ret, nil
+}
+
+func (o *oidSet) Release() error {
+	if o == nil || o.oidSet == nil {
+		return nil
+	}
+
+	var minor C.OM_uint32
+	cMajor := C.gss_release_oid_set(&minor, &o.oidSet)
+	if cMajor != C.GSS_S_COMPLETE {
+		return makeStatus(cMajor, minor)
+	}
+	o.pinner.Unpin()
+	return nil
+}

--- a/oidset_test.go
+++ b/oidset_test.go
@@ -1,0 +1,41 @@
+package gssapi
+
+import (
+	"testing"
+	"unsafe"
+
+	g "github.com/golang-auth/go-gssapi/v3"
+)
+
+func TestOidSet(t *testing.T) {
+	assert := NewAssert(t)
+
+	oids := []g.Oid{}
+
+	// test with no OIDs
+	oidSet, err := newOidSet(oids)
+	assert.NoError(err)
+	assert.NotNil(oidSet)
+
+	mechs := []g.GssMech{g.GSS_MECH_KRB5, g.GSS_MECH_IAKERB}
+
+	// Test with some OIDs
+	oids = append(oids, mechs[0].Oid(), mechs[1].Oid())
+	oidSet, err = newOidSet(oids)
+	assert.NoErrorFatal(err)
+	assert.NotNil(oidSet)
+
+	if oidSet != nil {
+		assert.Equal(2, int(oidSet.oidSet.count))
+
+		elms := oidSet.oidSet.elements
+		s := unsafe.Slice(elms, oidSet.oidSet.count)
+		assert.Equal(2, len(s))
+		for i, cOid := range s {
+			oid := oidFromGssOid(&cOid)
+			oidString, err := oid2String(oid)
+			assert.NoErrorFatal(err)
+			assert.Equal(mechs[i].OidString(), oidString)
+		}
+	}
+}

